src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,128 @@
+package com.pusher.client.channel.impl;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import com.google.gson.annotations.SerializedName;
+import com.pusher.client.AuthorizationFailureException;
+import com.pusher.client.Authorizer;
+import com.pusher.client.channel.ChannelState;
+import com.pusher.client.channel.PrivateEncryptedChannel;
+import com.pusher.client.channel.PrivateEncryptedChannelEventListener;
+import com.pusher.client.channel.SubscriptionEventListener;
+import com.pusher.client.connection.impl.InternalConnection;
+import com.pusher.client.crypto.nacl.SecretBoxOpener;
+import com.pusher.client.util.Factory;
+import com.pusher.client.util.internal.Base64;
+
+import java.lang.reflect.Type;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class PrivateEncryptedChannelImpl extends ChannelImpl implements PrivateEncryptedChannel {
+
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+    protected SecretBoxOpener secretBoxOpener;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    /* Base class overrides */
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+    protected byte[] checkAuthentication() {
```

Comment:
```suggestion
    private byte[] authenticate() {
```

The name change has been previously [requested](https://github.com/pusher/pusher-websocket-java/pull/234#discussion_r400098583).

When it comes to the visibility modifier, code should not be exposed just for tests as then it means that the tests test internal details (creating undesired coupling). The same can be tested by calling the `prepareSubscribeMessage` method which is provided to the client code ("to" -> "prepare" name change has been previously [requested](https://github.com/pusher/pusher-websocket-java/pull/234#discussion_r400046779)).

Reply:
Have updated the `checkAuthentication` and visibility. 

I'm not sure I want to change to the toSubscribeMessage at this time because it's implemented in all the other channels, but I take on board your point the naming isn't the clearest and have added to our future improvement doc.